src/crimson/os/seastore/extentmap_manager/btree/btree_extentmap_manager.cc

```diff
@@ -0,0 +1,126 @@
+  : tm(tm),
+    extmap_root(extmap_root) {}
+
+
+BtreeExtentMapManager::initialize_extmap_ret
+BtreeExtentMapManager::initialize_extmap(Transaction &t)
+{
+
+  logger().debug("BtreeExtentMapManager::initialize_extmap");
+  if (extmap_root.extmap_root_laddr != L_ADDR_NULL)
+    return initialize_extmap_ertr::make_ready_future<extmap_root_t>(extmap_root);
+
+  return tm.alloc_extent<ExtMapLeafNode>(t, L_ADDR_MIN, EXTMAP_BLOCK_SIZE)
+    .safe_then([this](auto&& root_extent) {
+      root_extent->set_size(0);
+      root_extent->set_depth(0);
+      extmap_root = extmap_root_t(0, root_extent->get_laddr());
+      return initialize_extmap_ertr::make_ready_future<extmap_root_t>(extmap_root);
+  });
+}
+
+BtreeExtentMapManager::get_root_ret
+BtreeExtentMapManager::get_extmap_root(Transaction &t)
+{
+  assert(extmap_root.extmap_root_laddr != L_ADDR_NULL);
+  laddr_t laddr = extmap_root.extmap_root_laddr;
+  return extmap_load_extent(tm, t, laddr, extmap_root.depth);
+}
+
+BtreeExtentMapManager::find_lextent_ret
+BtreeExtentMapManager::find_lextent(Transaction &t, objaddr_t lo,
+		                    extent_len_t len)
+{
+  logger().debug("BtreeExtentMapManager::find_lextent: {}, {}", lo, len);
+  return get_extmap_root(t).safe_then([this, &t, lo, len](auto&& extent) {
+    return extent->find_lextent(tm, t, lo, len);
+  }).safe_then([](auto &&e) {
+    logger().debug("BtreeExtentMapManager::find_lextent: found_lextent {}", e);
+    return find_lextent_ret(
+        find_lextent_ertr::ready_future_marker{},
+	std::move(e));
+  });
+
+}
+
+BtreeExtentMapManager::add_lextent_ret
+BtreeExtentMapManager::add_lextent(Transaction &t, objaddr_t lo,
+                                   lext_map_val_t val)
+{
+  assert(lo % PAGE_SIZE == 0 && val.length % PAGE_SIZE == 0);
+  logger().debug("BtreeExtentMapManager::add_lextent: {}, {}, {}", lo, val.laddr, val.length);
+  return get_extmap_root(t).safe_then([this, &t, lo, val](auto &&root) {
+    return insert_lextent(t, root, lo, val);
+  }).safe_then([](auto ret) {
+      logger().debug("BtreeExtentMapManager::add_lextent:  {}", ret);
+      return add_lextent_ret(
+        add_lextent_ertr::ready_future_marker{},
+        std::move(ret));
+  });
+
+}
+
+BtreeExtentMapManager::insert_lextent_ret
+BtreeExtentMapManager::insert_lextent(Transaction &t, ExtMapNodeRef root,
+                    objaddr_t logical_offset, lext_map_val_t val)
+{
+  auto split = insert_lextent_ertr::make_ready_future<ExtMapNodeRef>(root);
+  if (root->at_max_capacity()) {
+      logger().debug("BtreeExtentMapManager::splitting root {}", *root);
+    split =  root->extmap_alloc_extent<ExtMapInnerNode>(tm, t, EXTMAP_BLOCK_SIZE)
+        .safe_then([this, root, &t, logical_offset](auto&& nroot) {
+        nroot->set_depth(root->depth + 1);
+	nroot->journal_insert(nroot->begin(), OBJ_ADDR_MIN,
+			      root->get_laddr(), nullptr);
+        extmap_root.extmap_root_laddr = nroot->get_laddr();
+        extmap_root.depth = root->depth + 1;
```

Comment:
Is it appropriate to update the global `BtreeExtentMapManager::extmap_root` before `Transaction t` is submitted?

IIUC the transactions under SeaStore is fully isolated with each other, and can be failed due to potential conflicts. This leads to the following issues:
* the updated `extmap_root` with depth/laddr can be seen by other transactions, which is not valid before the transaction is submitted successfully;
* the updated `extmap_root` needs to be reverted if the transaction is failed;

Reply:
seems I should put it in a cachedExtent, then cache will handle it.

Reply:
seems I can't use cachedExtent for extmap root, it will be recorded in transaction and write to disk, but extmap root will be stored in onode.

Reply:
@tchaikov  and @athanatos @cyx1231st  any suggestion for solve it?  every time extentmap operation will transfer extmap root as input parameter and return extmap root val, so onode will updated extmap root info if it changed. and onodo is CachedExtent, so each transaction can get its own extmap root info according to onode update status.

Reply:
I think it is based on the decision whether to imbed the extmap root node or simply a laddr inside onode. I agree the transaction can work with updates correctly as long as the info is stored in the CachedExtent.

Reply:
If we assume that extmap_root is not global (and it won't be) but rather a sub-field of the onode (which will be contained within a CachedExtent), then the update to the onode will presumably include the extmap_root update in the same transaction.